File: typescript/src/events/event.ts
import { v4 as uuid } from 'uuid';
import { Buffer } from 'buffer';

export const ErrInvalidEvent = new Error('invalid event');
export const ErrPayloadSerialization = new Error('event.payload_serialization');
export const ErrInternal = new Error('internal event');

// Event
export class Event {
  private id: string;
  private entityId: string;
  private topic: string;
  private payload: Uint8Array;
  private timestamp: Date;

  constructor(
    id: string,
    entityId: string,
    topic: string,
    payload: Uint8Array,
    timestamp: Date,
  ) {
    if (id.length === 0) {
      throw ErrInvalidEvent;
    }

    if (entityId.length === 0) {
      throw ErrInvalidEvent;
    }

    if (topic.length === 0) {
      throw ErrInvalidEvent;
    }

    if (!payload || payload.length === 0) {
      throw ErrInvalidEvent;
    }

    this.id = id;
    this.entityId = entityId;
    this.topic = topic;
    this.payload = payload;
    this.timestamp = timestamp;
  }

  static create(entityId: string, topic: string, payload: unknown): Event {
    let buff;
    try {
      const json = JSON.stringify(payload);
      buff = Buffer.from(json);
    } catch (err) {
      throw ErrPayloadSerialization;
    }

    return new Event(uuid(), entityId, topic, new Uint8Array(buff), new Date());
  }

  getId(): string {
    return this.id;
  }

  getEntityId(): string {
    return this.entityId;
  }

  getTopic(): string {
    return this.topic;
  }

  getPayload(): Uint8Array {
    return this.payload;
  }

  deserializePayload<T>(): T {
    const buff = Buffer.from(this.payload);

    try {
      return JSON.parse(buff.toString());
    } catch (err) {
      throw ErrPayloadSerialization;
    }
  }

  getTimestamp(): Date {
    return this.timestamp;
  }
}

File: rust/src/events/event.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use thiserror::Error;
use uuid::Uuid;

#[derive(Error, Debug)]
pub enum EventError {
    #[error("invalid event")]
    Invalid,
    #[error("payload serialization")]
    PayloadSerialization,
    #[error("internal event error")]
    Internal,
}

// Event
#[derive(Debug, Clone, PartialEq)]
pub struct Event {
    id: String,
    entity_id: String,
    topic: String,
    payload: Vec<u8>,
    timestamp: DateTime<Utc>,
}

impl Event {
    pub fn new(
        id: String,
        entity_id: String,
        topic: String,
        payload: Vec<u8>,
        timestamp: DateTime<Utc>,
    ) -> Result<Event, EventError> {
        if id.is_empty() {
            return Err(EventError::Invalid);
        }

        if entity_id.is_empty() {
            return Err(EventError::Invalid);
        }

        if topic.is_empty() {
            return Err(EventError::Invalid);
        }

        if payload.is_empty() {
            return Err(EventError::Invalid);
        }

        Ok(Event {
            id,
            entity_id,
            topic,
            payload,
            timestamp,
        })
    }

    pub fn create<I, T, P>(entity_id: I, topic: T, payload: &P) -> Result<Event, EventError>
    where
        I: Into<String>,
        T: Into<String>,
        P: Serialize,
    {
        let entity_id = entity_id.into();
        let topic = topic.into();

        let payload = serde_json::to_vec(payload).map_err(|_| EventError::PayloadSerialization)?;

        Event::new(
            Uuid::new_v4().to_string(),
            entity_id,
            topic,
            payload,
            Utc::now(),
        )
    }

    pub fn id(&self) -> &str {
        &self.id
    }

    pub fn entity_id(&self) -> &str {
        &self.entity_id
    }

    pub fn topic(&self) -> &str {
        &self.topic
    }

    pub fn payload(&self) -> &[u8] {
        &self.payload
    }

    pub fn deserialize_payload<'a, T>(&'a self) -> Result<T, EventError>
    where
        T: Deserialize<'a>,
    {
        serde_json::from_slice(&self.payload).map_err(|_| EventError::PayloadSerialization)
    }

    pub fn timestamp(&self) -> &DateTime<Utc> {
        &self.timestamp
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(Debug, Serialize, Deserialize)]
    struct Data {
        msg: String,
    }

    #[test]
    fn create() {
        let res = Event::create(
            "entity#01",
            "topic.code",
            &Data {
                msg: "Hello World".to_string(),
            },
        );
        assert!(res.is_ok());

        let event = res.unwrap();
        assert_eq!(event.entity_id(), "entity#01");
        assert_eq!(event.topic(), "topic.code");
        assert!(event.payload().len() > 0);
    }

    #[test]
    fn payload_serialization_and_deserialization() {
        let event = Event::create(
            "entity#01",
            "topic.code",
            &Data {
                msg: "Hello World".to_string(),
            },
        )
        .unwrap();

        let res = event.deserialize_payload();
        assert!(res.is_ok());

        let data: Data = res.unwrap();
        assert_eq!(data.msg, "Hello World");
    }
}
